### DISCUSSION TITLE
Performance gap with PlayCanvas on MacOS

### DISCUSSION THREAD

**Discussion by xiasun** [mkkellogg/GaussianSplats3D#301] [Ideas]
I have gone through several web implementations including this Three.js one, the Babylon embedded one, and Gsplat.js. All of these share similar performance on windows PC with the same data. However, the [PlayCanvas one](https://playcanvas.com/supersplat/editor/) is significantly faster on my MacOS. With 5 million splats, the framerate is about 3 to 4 times higher. I wonder if you have any suggestions for this?

My major work are based on Babylon so I start a [thread](https://forum.babylonjs.com/t/possible-performance-bottleneck-on-high-resolution-devices-macos/52660) there to describe what I have done so far. But I thought this problem might also be a common issue with this Three.js impls.

I would appreciate any suggestions you might have for me.

**Comment by mkkellogg** [comment]
Do you have a specific scene you could share that exhibits the issue? Maybe i can use that to start debugging. Also, could you share your code for initializing the `Viewer`?

**Reply by xiasun** [reply]
I eventually found out the huge performance gap between SuperSplat (PlayCanvas) and other WebGL implementations was a bug of chromium on MacOS, that render a same scene with same params to texture was FASTER that render to back buffer directly. It has nothing to do with GS itself. Thanks!

**Comment by d0rianb** [comment]
I have looked in PlayCanvas source code, and specifically [this pull request ](https://github.com/playcanvas/engine/pull/6357/) which seems to have well improved their splat rendering performances.  
It [looks like](https://github.com/playcanvas/supersplat/blob/79928b11b43e8ff88adb5eb452fede8d63920e51/src/splat-convert.ts#L392) they are using Morton encoding to sort their splat, and that they are submitting it to the GPU directly via a Texture, instead of a FloatArray / IntBuffer.
On large scene (>1M splat), it seems to make a difference, at least on my Macbook.

**Reply by mkkellogg** [reply]
In the current code, if you don't load scene progressively, my viewer makes similar optimizations: It re-orders the splats according to spatial proximity. Additionally my viewer has always stored splat data in textures 😄

**Reply by d0rianb** [reply]
Ok thanks, I didn't noticed that disabling progressive loading would have such an impact on performance improvement :)
One more thing, why is the sorting only partially done on the GPU ? Is there a limitation that prevents this, or is it something left to do ?

**Reply by mkkellogg** [reply]
Well by default the sorting is done completely on the CPU. When `gpuAcceleratedSort` is set to `true`, you're right about the sort only being partially done on the GPU (just the splat distances are computed on the GPU). This is because the viewer is based on WebGL, and there's not a good way to do a parallelized sort on the GPU with WebGL. It is possible to pre-compute the splat distances are on the GPU though, because that operation is easily parallelizable.

**Reply by xiasun** [reply]
@d0rianb The [motion order](https://forum.playcanvas.com/t/gaussian-splatting-playcanvas/33503/67) gave me a small improvement (fps 18 to 23). If you're on an Apple Silicon Mac, I guess there would be no more big problem then. On Intel Mac, there will be an additional [RT issue](https://forum.playcanvas.com/t/gaussian-splatting-playcanvas/33503/72) which has a 4 times performance gap.

**Reply by nancywan1004** [reply]
I'm also curious about the perf for large scenes (>1M splats). I toggled off the progressive load and tried with the Edge browser on a  Silicon Macbook pro. The perf is only about 1/3 of which on playcanvas. I also tried with turning on the gpuAcceleratedSort but looks like there's a bug since nothing would load and no error... Is there any good idea on optimizations in this type of scenario? Thanks!

**Reply by xiasun** [reply]
@nancywan1004 On Intel Macs, I found that rendering to a render target is four times faster than rendering directly to the canvas, which creates a performance gap between this project and PlayCanvas. You can review this [issue](https://issuetracker.google.com/issues/360169322) and compare the frame rates between the two snippets: [Render to canvas directly](https://playground.babylonjs.com/#8Y719R#19) and [Render to texture](https://playground.babylonjs.com/#8Y719R#22). If you experience the same problem on your machine, I recommend trying to render on a [WebGLRenderTarget](https://threejs.org/docs/#api/en/renderers/WebGLRenderTarget) to bypass the issue.

P.S.
You can increase the value of `numInstances` in the snippets to match your own dataset. The value `numInstances * numQuadsPerInstance` should equal the total count of your point cloud.
```
const numInstances = 10000;
const numQuadsPerInstance = 128;
```
